File: front-end/src/app/domain/models/game/action/impl/gamePlay/DeclareBankruptcyAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";
import {GameAssetBid} from "../../../GameAssetBid";
import {OwnableLocation} from "../../../board/location/OwnableLocation";
import {JailBreakCard} from "../../../card/impl/JailBreakCard";
import {GameAssetBidUtil} from "../../../GameAssetBidUtil";
import {LocationUtil} from "../../../board/location/LocationUtil";
import {CardUtil} from "../../../card/CardUtil";

/**
 * TODO: Add JsDoc
 *
 * @author Hamza el Haouti
 */
export class DeclareBankruptcyAction extends GameAction {

  public constructor(
    id: number,
    actor: Player,
    performedAt: Date,
    removedBids: GameAssetBid[],
    removedOwnedLocations: OwnableLocation[],
    removedOwnedJailBreakCard: JailBreakCard[],
    removedMoney: number,
    removedPosition: number
  ) {
    super(DeclareBankruptcyAction.name, id, actor, performedAt);
    this._removedBids = removedBids;
    this._removedOwnedLocations = removedOwnedLocations;
    this._removedOwnedJailBreakCard = removedOwnedJailBreakCard;
    this._removedMoney = removedMoney;
    this._removedPosition = removedPosition;
  }

  private _removedBids: GameAssetBid[];

  get removedBids(): GameAssetBid[] {
    return this._removedBids;
  }

  private _removedOwnedLocations: OwnableLocation[];

  get removedOwnedLocations(): OwnableLocation[] {
    return this._removedOwnedLocations;
  }

  private _removedOwnedJailBreakCard: JailBreakCard[];

  get removedOwnedJailBreakCard(): JailBreakCard[] {
    return this._removedOwnedJailBreakCard;
  }

  private _removedMoney: number;

  get removedMoney(): number {
    return this._removedMoney;
  }

  private _removedPosition: number;

  get removedPosition(): number {
    return this._removedPosition;
  }

  public static trueCopy(action: DeclareBankruptcyAction): GameAction | null {
    return new DeclareBankruptcyAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.removedBids.map(GameAssetBidUtil.trueCopy),
      action.removedOwnedLocations.map(loc => <OwnableLocation>LocationUtil.trueCopy(loc)),
      action.removedOwnedJailBreakCard.map(card => <JailBreakCard>CardUtil.trueCopy(card)),
      action.removedMoney,
      action.removedPosition
    );
  }

  public override toString(): string {
    return `${this.actor.user.username} was declared bankrupt.`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/gamePlay/DeclareBankruptcyAction.java
package nl.hva.backend.domain.models.game.action.impl.gamePlay;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.GameAsset;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.GameAction;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;
import nl.hva.backend.domain.models.game.card.impl.JailBreakCard;

import javax.persistence.Entity;
import javax.persistence.ManyToMany;
import java.util.List;

/**
 * Representation, of the action, of declaring a player bankrupt.
 * <p>
 * When a player is declared bankrupt all belongings and bid history of the player are removed. OwnableLocations are
 * then again put up for sale by the bank.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class DeclareBankruptcyAction extends GameAction {

    @ManyToMany
    private List<GameAsset.GameAssetBid> removedBids;

    @ManyToMany
    private List<OwnableLocation> removedOwnedLocations;

    @ManyToMany
    private List<JailBreakCard> removedOwnedJailBreakCard;

    private double removedMoney;

    private int removedPosition;

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param actor The player performing this action.
     */
    public DeclareBankruptcyAction(
            Player actor,
            List<GameAsset.GameAssetBid> removedBids,
            List<OwnableLocation> removedOwnedLocations,
            List<JailBreakCard> removedOwnedJailBreakCard,
            double removedMoney,
            int removedPosition
    ) {
        super(actor);
        this.removedBids = removedBids;
        this.removedOwnedLocations = removedOwnedLocations;
        this.removedOwnedJailBreakCard = removedOwnedJailBreakCard;
        this.removedMoney = removedMoney;
        this.removedPosition = removedPosition;
    }

}